zerver/lib/alert_words.py

```diff
@@ -2,81 +2,121 @@
 
 import ahocorasick
 from django.db import transaction
+from pydantic import BaseModel, ConfigDict, StringConstraints
+from typing_extensions import Annotated
 
 from zerver.lib.cache import (
     cache_with_key,
-    realm_alert_words_automaton_cache_key,
-    realm_alert_words_cache_key,
+    realm_watched_phrases_automaton_cache_key,
+    realm_watched_phrases_cache_key,
 )
 from zerver.models import AlertWord, Realm, UserProfile
-from zerver.models.alert_words import flush_realm_alert_words
+from zerver.models.alert_words import flush_realm_watched_phrases
 
 
-@cache_with_key(lambda realm: realm_alert_words_cache_key(realm.id), timeout=3600 * 24)
-def alert_words_in_realm(realm: Realm) -> Dict[int, List[str]]:
-    user_ids_and_words = AlertWord.objects.filter(realm=realm, user_profile__is_active=True).values(
-        "user_profile_id", "word"
-    )
-    user_ids_with_words: Dict[int, List[str]] = {}
-    for id_and_word in user_ids_and_words:
-        user_ids_with_words.setdefault(id_and_word["user_profile_id"], [])
-        user_ids_with_words[id_and_word["user_profile_id"]].append(id_and_word["word"])
-    return user_ids_with_words
-
-
-@cache_with_key(lambda realm: realm_alert_words_automaton_cache_key(realm.id), timeout=3600 * 24)
-def get_alert_word_automaton(realm: Realm) -> ahocorasick.Automaton:
-    user_id_with_words = alert_words_in_realm(realm)
-    alert_word_automaton = ahocorasick.Automaton()
-    for user_id, alert_words in user_id_with_words.items():
-        for alert_word in alert_words:
-            alert_word_lower = alert_word.lower()
-            if alert_word_automaton.exists(alert_word_lower):
-                (key, user_ids_for_alert_word) = alert_word_automaton.get(alert_word_lower)
-                user_ids_for_alert_word.add(user_id)
+class WatchedPhraseData(BaseModel):
+    model_config = ConfigDict(extra="forbid", frozen=True)
+
+    watched_phrase: Annotated[str, StringConstraints(max_length=100)]
```

Comment:
I think it'd be better to name this simply `phrase`, so we can simply refer to it as `WatchedPhraseData.phrase`.

and in future.. `WatchedPhraseData.automatically_follow_topics`